SUPESC-606 Updated Availability Notification IG with stores to Yves host mapping configuration. (#1629)
spryker/spryker-docs: _includes/pbc/all/install-features/202304.0/install-the-availability-notification-feature.md

| AvailabilityNotificationSubscriber | This plugin is responsible for listening to and processing product availability changes. |               | Spryker\Zed\AvailabilityNotification\Business\Subscription |

**src/Pyz/Zed/Event/EventDependencyProvider.php**

```php
<?php

namespace Pyz\Zed\Event;

use Spryker\Zed\AvailabilityNotification\Communication\Plugin\Event\Subscriber\AvailabilityNotificationSubscriber;

use Spryker\Zed\Event\EventDependencyProvider as SprykerEventDependencyProvider;

class EventDependencyProvider extends SprykerEventDependencyProvider
{
     /**
     * @return \Spryker\Zed\Event\Dependency\EventSubscriberCollectionInterface
     */
     public function getEventSubscriberCollection()
     {
         $eventSubscriberCollection = parent::getEventSubscriberCollection();
        
         $eventSubscriberCollection->add(new AvailabilityNotificationSubscriber());
        
         return $eventSubscriberCollection;
     }
}
```

{% info_block warningBox "Verification" %}

To verify that `AvailabilityNotificationSubscriber` is working, follow these steps:

1. Add a new product and make it unavailable.

> 2. As a customer, subscribe to its availability notifications on Yves.

Author: This is a part of the original IG and I guess it means Storefront.

```php
<?php

namespace Pyz\Zed\Customer;

use Spryker\Zed\AvailabilityNotification\Communication\Plugin\Customer\AvailabilityNotificationSubscriptionCustomerTransferExpanderPlugin;
use Spryker\Zed\AvailabilityNotification\Communication\Plugin\CustomerAnonymizer\AvailabilityNotificationAnonymizerPlugin;
use Spryker\Zed\Customer\CustomerDependencyProvider as SprykerCustomerDependencyProvider;

class CustomerDependencyProvider extends SprykerCustomerDependencyProvider
{
    ...

     /**
     * @return \Spryker\Zed\Customer\Dependency\Plugin\CustomerAnonymizerPluginInterface[]
     */
     protected function getCustomerAnonymizerPlugins()
     {
         return [
             ...
             new AvailabilityNotificationAnonymizerPlugin(),
         ];
     }

     /**
     * @return \Spryker\Zed\Customer\Dependency\Plugin\CustomerTransferExpanderPluginInterface[]
     */
     protected function getCustomerTransferExpanderPlugins()
     {
         return [
         ...
            new AvailabilityNotificationSubscriptionCustomerTransferExpanderPlugin(),
         ];
     }
}
```
</details>

{% info_block warningBox "Verification" %}

> To verify that `AvailabilityNotificationAnonymizerPlugin` is working:

Reviewer: ```suggestion
To verify that `AvailabilityNotificationAnonymizerPlugin` works:
```

1. Add a new product.
2. On the Storefront, as a company user, subscribe to its availability notifications.

> 3. Check that the corresponding line is added to the `spy_availability_notification_subscription` table.</li><li>Delete this user.

Reviewer: rewrite the sentence without </li><li>
```suggestion
3. Check that the corresponding line is added to the `spy_availability_notification_subscription` table.</li><li>Delete this user.
```

4. Check that the line is deleted from the `spy_availability_notification_subscription` table.

> To verify that `AvailabilityNotificationSubscriptionCustomerTransferExpanderPlugin` is working:

Reviewer: ```suggestion
To verify that `AvailabilityNotificationSubscriptionCustomerTransferExpanderPlugin` works:
```

1. Add a new product.
2. On the Storefront, as a company user, subscribe to its availability notifications.
3. On the Storefront, go to account overview > *Newsletters*.
4. Check that you are subscribed to the product's availability notifications.

{% endinfo_block %}

### 4) Set up the configuration

{% info_block infoBox %}

> You can define stores to Yves hosts mapping. This allows generating correct URLs (e.g. unsubscription URL) for multi-store setup.

Reviewer 2: probably store sharing should be mentioned

| AvailabilityNotificationSubscriptionResponseTransfer      | class  | created | src/Generated/Shared/Transfer/AvailabilityNotificationSubscriptionResponseTransfer.php |
| AvailabilityNotificationSubscriptionRequestTransfer       | class  | created | src/Generated/Shared/Transfer/AvailabilityNotificationSubscriptionRequestTransfer.php  |
| AvailabilityNotificationSubscriptionMailDataTransfer      | class  | created | src/Generated/Shared/Transfer/AvailabilityNotificationSubscriptionMailDataTransfer.php |
| AvailabilityNotificationDataTransfer                      | class  | created | src/Generated/Shared/Transfer/AvailabilityNotificationDataTransfer.php                 |
| CustomerTransfer.availabilityNotificationSubscriptionSkus | column | created | src/Generated/Shared/Transfer/CustomerTransfer.php                                     |
| MailTransfer.availabilityNotificationSubscriptionMailData | column | created | src/Generated/Shared/Transfer/MailTransfer.php                                         |

{% endinfo_block %}

### 3) Set up behavior

Enable the following behaviors by registering the plugins.

> #### Listening to the availability_notification event

Reviewer: Rewrite the title. It should be a task-based heading and start with a bare infinitive. For example:
```suggestion
#### Configure listening to the availability_notification event
```

| AvailabilityNotificationSubscriber | This plugin is responsible for listening to and processing product availability changes. |               | Spryker\Zed\AvailabilityNotification\Business\Subscription |

**src/Pyz/Zed/Event/EventDependencyProvider.php**

```php
<?php

namespace Pyz\Zed\Event;

use Spryker\Zed\AvailabilityNotification\Communication\Plugin\Event\Subscriber\AvailabilityNotificationSubscriber;

use Spryker\Zed\Event\EventDependencyProvider as SprykerEventDependencyProvider;

class EventDependencyProvider extends SprykerEventDependencyProvider
{
     /**
     * @return \Spryker\Zed\Event\Dependency\EventSubscriberCollectionInterface
     */
     public function getEventSubscriberCollection()
     {
         $eventSubscriberCollection = parent::getEventSubscriberCollection();
        
         $eventSubscriberCollection->add(new AvailabilityNotificationSubscriber());
        
         return $eventSubscriberCollection;
     }
}
```

{% info_block warningBox "Verification" %}

> To verify that `AvailabilityNotificationSubscriber` is working, follow these steps:

Reviewer: ```suggestion
To verify that `AvailabilityNotificationSubscriber` works, follow these steps:
```

1. Add a new product and make it unavailable.
2. As a customer, subscribe to its availability notifications on the Storefront.
3. Make the product available.
4. Check your mailbox for the email about the product's availability.

{% endinfo_block %}

> #### Email handling

Reviewer: The same. Rephrase it. For example:
```suggestion
#### Configure email handling
```

     {
         $container = parent::provideBusinessLayerDependencies($container);
        
         $container->extend(static::MAIL_TYPE_COLLECTION, function (MailTypeCollectionAddInterface $mailCollection) {
             $mailCollection
             ...
             ->add(new AvailabilityNotificationUnsubscribedMailTypePlugin())
             ->add(new AvailabilityNotificationSubscriptionMailTypePlugin())
             ->add(new AvailabilityNotificationMailTypePlugin())
             ...
             ;
    
            return $mailCollection;
        });
        ...

        return $container;
    }
}
```
</details>

{% info_block warningBox "Verification" %}

Verify that `AvailabilityNotificationSubscriptionMailTypePlugin`, `AvailabilityNotificationUnsubscribedMailTypePlugin`, and `AvailabilityNotificationMailTypePlugin` work:
1. Add a new product.
2. On the Storefront, as a customer, subscribe to its availability notifications.
3. Switch the availability status of the product several times.
4. Check your mailbox for emails about the product's status being switched to available and unavailable.

{% endinfo_block %}

> #### Customer behavior

Reviewer: the same. Rephrase it. For example:
```suggestion
#### Configure customer behavior
```

```php
<?php

namespace Pyz\Zed\Customer;

use Spryker\Zed\AvailabilityNotification\Communication\Plugin\Customer\AvailabilityNotificationSubscriptionCustomerTransferExpanderPlugin;
use Spryker\Zed\AvailabilityNotification\Communication\Plugin\CustomerAnonymizer\AvailabilityNotificationAnonymizerPlugin;
use Spryker\Zed\Customer\CustomerDependencyProvider as SprykerCustomerDependencyProvider;

class CustomerDependencyProvider extends SprykerCustomerDependencyProvider
{
    ...

     /**
     * @return \Spryker\Zed\Customer\Dependency\Plugin\CustomerAnonymizerPluginInterface[]
     */
     protected function getCustomerAnonymizerPlugins()
     {
         return [
             ...
             new AvailabilityNotificationAnonymizerPlugin(),
         ];
     }

     /**
     * @return \Spryker\Zed\Customer\Dependency\Plugin\CustomerTransferExpanderPluginInterface[]
     */
     protected function getCustomerTransferExpanderPlugins()
     {
         return [
         ...
            new AvailabilityNotificationSubscriptionCustomerTransferExpanderPlugin(),
         ];
     }
}
```
</details>

{% info_block warningBox "Verification" %}

> To verify that `AvailabilityNotificationAnonymizerPlugin` works:

Reviewer: ```suggestion
Verify that `AvailabilityNotificationAnonymizerPlugin` works:
```

1. Add a new product.
2. On the Storefront, as a company user, subscribe to its availability notifications.
3. Check that the corresponding line is added to the `spy_availability_notification_subscription` table.
4. Delete the company user that is subscribed to the product's availability.
5. Check that the line is deleted from the `spy_availability_notification_subscription` table.

> To verify that `AvailabilityNotificationSubscriptionCustomerTransferExpanderPlugin` works:

Reviewer: ```suggestion
Verify that `AvailabilityNotificationSubscriptionCustomerTransferExpanderPlugin` works:
```

1. Add a new product.
2. On the Storefront, as a company user, subscribe to its availability notifications.

> 3. On the Storefront, go to account overview > *Newsletters*.

Reviewer: ```suggestion
3. On the Storefront, go to account overview&nbsp;<span aria-label="and then">></span> **Newsletters**.
```
